macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala

```diff
@@ -0,0 +1,46 @@
+package com.softwaremill.macwire.internals
+
+import scala.reflect.macros.blackbox
+
+private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger) {
+  import c.universe._
+
+  lazy val dependencyResolver = new DependencyResolver[c.type](c, log)
+
+  lazy val targetType: Type = implicitly[c.WeakTypeTag[T]].tpe
+
+  lazy val companionType: Option[Type] = if(targetType.companion == NoType) None else Some(targetType.companion)
+
+  def isCompanionApply(method: Symbol): Boolean =
+    method.isMethod &&
+    method.isPublic &&
+    method.asMethod.returnType <:< targetType &&
+    method.asMethod.name.decodedName.toString == "apply"
+
+  lazy val applies: Option[List[Symbol]] = log.withBlock("Looking for apply methods of Companion Object") {
+    val as: Option[List[Symbol]] = companionType.map(_.members.filter(isCompanionApply).toList)
+    as.foreach(x => log.withBlock(s"There are ${x.size} apply methods:" ) { x.foreach(c => log(showApply(c))) })
+    as
+  }
+
+  lazy val apply: Option[Symbol] = applies.flatMap( _ match {
+    case applyMethod :: Nil => Some(applyMethod)
+    case _ => None
+  })
+
+  lazy val applySelect: Option[Select] = apply.map(a => Select(Ident(targetType.typeSymbol.companion), a))
```

Comment:
I don't quite understand the idea with all these `lazy val`s :) Why are they lazy and public? They make the code quite hard to follow ... there's a lot of jumping around. Maybe this could be simplified, and when possible made private?